### ISSUE TITLE
Save and load in binary, compatible with NumPy/Matlab and others

### ISSUE THREAD

**Issue by certik** [fortran-lang/stdlib#486]
* NumPy: https://numpy.org/devdocs/reference/generated/numpy.lib.format.html
* Matlab: https://www.mathworks.com/help/matlab/import_export/mat-file-versions.html
* SciPy code to load/write MAT files: https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.loadmat.html
* Julia library to read/write MAT files: https://github.com/JuliaIO/MAT.jl

First requested [here](https://fortran-lang.discourse.group/t/what-are-your-favorite-and-least-favorite-things-about-fortran/27/49).

**Comment by milancurcic** [comment]
This seems useful and in scope. I often work with MAT files (of various versions) from colleagues and I use SciPy.io `loadmat` and `savemat`.

For my own interoperable binary data between Fortran and Python, I use NetCDF. I don't think any of the language-specific binary formats will beat it in terms of features, performance, or stability. Likewise for HDF5 which is suitable for unstructured data.

**Comment by MarDiehl** [comment]
There is already an HDF5 writer/reader which looks promising: https://github.com/geospace-code/h5fortran. I uses the Fortran bindings of the C library.
I think it is reasonable to keep HDF5 support out of stdlib, it is neither part of the C nor the python standard library.

**Comment by awvwgk** [comment]
I got the basic structure for reading and writing npy files implemented in https://github.com/fortran-lang/stdlib/pull/581. Needs some polishing, especially the reading, and much more unit tests to cover all possible errors the loading can encounter.

**Comment by TejasAvinashShetty** [comment]
[libnpy](http://www.maths.unsw.edu.au/~mclean/libnpy-0.5.tgz) seems to be a  library that provides simple routines for saving a C or Fortran array to a data file using NumPy's own binary format. 
Please see https://scipy-cookbook.readthedocs.io/items/InputOutput.html
> Not my idea  See first [CAZT's](https://stackoverflow.com/users/8644947/cazt) comment on [CAZT's  stackoverflow answer](https://stackoverflow.com/a/46333419/7952027)

**Comment by jvdp1** [comment]
> There is already an HDF5 writer/reader which looks promising: https://github.com/geospace-code/h5fortran. I uses the Fortran bindings of the C library. I think it is reasonable to keep HDF5 support out of stdlib, it is neither part of the C nor the python standard library.

I agree with @MarDiehl . I recently used @scivision 's `h5fortran` and found it great and really easy to use. Therefore, I also think reasonable to keep HDF5 support out of stdlib for the moment.

**Comment by awvwgk** [comment]
How do we want to handle the *npz* format? It is a *zip* archive with *npy* files. Probably, we have to develop a general interface for interacting with compressed archives first.

For the *mat* format I found a specification of the layout (linked in description at the top), should be straight-forward to code up, but I don't think I have a matlab version I could use to verify it, but I could try SciPy.

**Comment by ivan-pi** [comment]
> For the _mat_ format I found a specification of the layout (linked in description at the top), should be straight-forward to code up, but I don't think I have a matlab version I could use to verify it, but I could try SciPy.

Was your idea to implement the reader/writer entirely in Fortran based upon the PDF document, or call into the [MATLAB C API to Read MAT-File Data](https://de.mathworks.com/help/matlab/matlab-c-api-to-read-mat-file-data.html?s_tid=CRUX_lftnav)? The latter requires the client has the `libmat` shared run-time library located in `matlabroot/bin/arch`.

**Comment by awvwgk** [comment]
I was reading the specs, sounds easy enough to implement this from scratch and verify using SciPy. Unfortunately, the data can be compressed, and we need an interface to zlib or similar first.

Having the possibility to dynamically load a library with dlopen in case the matlab runtime libraries are around would be another option. However, than we first need an interface for dynamic loading.

**Comment by arjenmarkus** [comment]
My dynlib module in
https://sourceforge.net/p/flibs/svncode/HEAD/tree/trunk/src/dynlib/ could
serve as a starting point.

Op zo 12 dec. 2021 om 17:03 schreef Sebastian Ehlert <
***@***.***>:

**Comment by adenchfi** [comment]
I see stdlib has save_npy and load_npy functionality! I tested it out and it works great! I was wondering, if possible, if dim(:,:,:,:) arrays could also be supported. I only see interfaces up to rank-3.

**Comment by awvwgk** [comment]
They should be supported up to the maximum rank stdlib was configured for. The docs are only generated up to rank 3 to save space, while the fpm version allow up to rank 4, the CMake version can go up to rank 15.

**Comment by adenchfi** [comment]
Oh, thanks! I should have tested first, I took the docs too literally.

**Comment by ivan-pi** [comment]
While scrolling through the ARCHER2 super-computing service documentation I learned there is BSD-licensed library for MATLAB MAT files called [`matio`](https://github.com/tbeu/matio). It also has a [Fortran interface](https://github.com/tbeu/matio/tree/master/src/fortran) (help wanted https://github.com/tbeu/matio/issues/51), however it is doesn't appear to use the standard Fortran/C interoperability.

As @awvwgk has remarked above, supporting MATLAB binary files would require a [zlib](https://www.zlib.net/) interface and potentially also HDF5, both of which are available as C libraries. It looks more straightforward to just have a thin Fortran wrapper of a C/C++ implementation, than to write an interface/implementation for zlib (and HDF5) first.

**Comment by ivan-pi** [comment]
Since Fortran doesn't have positional or keyword arguments in the way Python does, for `.npz` files it seems more natural to adopt an API similar to the one in [NPY for Fortran](https://github.com/MRedies/NPY-for-Fortran):

```fortran
subroutine add_npz(zipfile,var_name,array)
   character(len=*), intent(in) :: zipfile
   character(len=*), intent(in) :: var_name
   real|complex|integer, intent(in) :: array(..)
```

Alternatively, we could have a handle based approach:

```fortran
integer :: npz_unit
real :: A(2,2)
complex :: B(3,3)

call open_npz(newunit=npz_unit,filename="foo.npz")
call stage_npz(npz_unit,A,"A")
call stage_npz(npz_unit,B,"B")
call close_npz(npz_unit)
```

Since Fortran uses integer units as file handles, the concept should be familiar already.

**Comment by ivan-pi** [comment]
The `.npz` format is also useful to read Scipy sparse matrix formats (CSC, CSR, BSR, DIA, COO). See [`scipy.sparse.save_npz`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.save_npz.html#scipy-sparse-save-npz) for a description. The implementation can be found [here](https://github.com/scipy/scipy/blob/v1.8.1/scipy/sparse/_matrix_io.py#L11-L72). Note the keywords in the dictionary creation specify the array names.
